### ISSUE TITLE
"Basic Usage" documentation doesn't describe how to use the virtual environment

### ISSUE THREAD

**Issue by nicholasbishop** [pypa/pipenv#1144]
The [Basic Usage of Pipenv](https://docs.pipenv.org/basics.html) document only covers how to create and modify the virtual environment, but nothing about how to run it. The `pipenv run` command is not mentioned, and `pipenv shell` is only mentioned in terms of configuration, not usage.

I wanted to find this information because I was told in [another bug](https://github.com/kennethreitz/pipenv/issues/989#issuecomment-340237102) that "You really shouldn't be doing much of anything from inside a pipenv shell. This is a convenience function for people who really need subshells." If I understand correctly that means I should always be using `pipenv run` instead of `pipenv shell`. Unfortunately neither `pipenv run -h` nor `pipenv run --help` tells you how to use the `run` command, which is rather frustrating. And other commands such as `pipenv install` tell you to activate the `pipenv shell` at the end of the output, which makes it sound like the advice I quoted above is just wrong.

All of which is to say that I'm really confused and having more info in the "Basic Usage" document would be super helpful :)

**Comment by MatMoore** [comment]
I'm new to using pipenv, but I think it's misleading to say you shouldn't be doing much of anything from pipenv shell. Why not? This is the same workflow as virtualenv has, so it seems natural to work that way.

If the point is that I shouldn't be calling `pipenv`from inside a `pipenv shell`, that's also quite surprising. Again, coming from virtualenv I'm used to running `pip install` inside a virtualenv, so I'd expect to be able to use `pipenv install` within `pipenv shell`. When you call `pipenv shell` inside another shell, you get an error message saying not to do that, but there is no warnings when you use other subcommands.

IMO `pipenv shell` is more user friendly than typing `pipenv run` in front of every command. Setting up an environment is something I want to do once, and then not have to think about when I'm not actively installing/updating/removing dependencies.

I agree the docs could make the intended workflow clearer:

- The video embedded in https://docs.pipenv.org/index.html promotes using `pipenv shell`
- https://docs.pipenv.org/#other-commands explains what `shell` and `run` do but doesn't suggest any particular workflow
- https://docs.pipenv.org/basics.html#about-shell-configuration refers to using `shell` but doesn't explain what it is

**Comment by kennethreitz** [comment]
@nicholasbishop send pull requests!

**Comment by nicholasbishop** [comment]
Hi, it's not obvious to me what closing this issue means. Is it a valid issue, and therefore a pull request would be welcome? Or is it considered closed/invalid and so a PR is not helpful?

Github issues don't provide anything in the way of semantics around closing an issue, so a word or two in a comment to give guidance would be very helpful. Thanks :)

**Comment by kennethreitz** [comment]
If you think the documentation could be improved, send pull requests to improve it! Pretty easy :)

Issues aren't helpful for this kind of stuff.

**Comment by nicholasbishop** [comment]
I was thinking the issue might be helpful as a place to discuss what the "right" documentation would be. @MatMoore's comment above, for example, looks really well-thought out and a useful contribution to the discussion. I'm a relative newbie to pipenv, so I don't think it's quite fair to just say "put up a PR" if I don't even know what should go in it.

**Comment by techalchemy** [comment]
@nicholasbishop you're not being a pain in the ass, we are definitely behind on documentation and with holidays coming up many of us are currently getting killed with work deadlines (I know I am).  Pipenv shells are generally fine, but they probably shouldn't be used in production.  The reason there isn't much documentation for `pipenv run` is just because there isn't much to document so it probably never occurred to us/we never got around to it.

The command consumes everything after it and acts like you ran it in an activated virtualenv for your project.  This basically just means you never have to activate virtual environments in order to run code, so you don't have to worry about which virtual environment you are in, whether you deactivated, etc.  This is the main reason why pipenv uses the project name + a hash of its absolute path as a deterministic name for the virtualenv it creates, so that when you run `pipenv run` inside the directory structure pipenv knows exactly where to find the relevant virtualenv and execute in that context.

Basically, if you call `pipenv run somescript.py` it is similar to calling `/path/to/virtualenv/python somescript.py`, except pipenv provides additional convenience such as loading your `.env` file into the environment and such.  This way you can context switch or directory switch or, in your editor, instead of constantly sourcing virtualenvs, prefix all your commands with `pipenv run` and it will figure things out for you.  It's not that you're not _supposed_ to do things inside a pipenv shell, because really, its fine.  Kenneth uses the fish plugin for pipenv shells, so I'm pretty sure he makes heavy use of that feature.  It's really just that you lose some of the benefits when you do that.
